tokio-util/src/net/unix/mod.rs

```diff
@@ -0,0 +1,108 @@
+//! Unix domain socket helpers.
+
+use std::future::Future;
+use std::io::Result;
+use std::pin::Pin;
+
+/// Listening address.
+#[derive(Debug)]
+pub enum ListenAddr {
+    /// Socket address.
+    SocketAddr(std::net::SocketAddr),
+    /// Unix socket.
+    UnixSocket(tokio::net::unix::SocketAddr),
+}
+
+impl From<std::net::SocketAddr> for ListenAddr {
+    fn from(addr: std::net::SocketAddr) -> Self {
+        Self::SocketAddr(addr)
+    }
+}
+
+impl From<tokio::net::unix::SocketAddr> for ListenAddr {
+    fn from(addr: tokio::net::unix::SocketAddr) -> Self {
+        Self::UnixSocket(addr)
+    }
+}
+
+/// A trait for a listener: `TcpListener` and `UnixListener`.
+pub trait Listener: Send + Unpin {
+    /// Accepts a new incoming connection from this listener.
+    fn accept<'a>(
+        &'a self,
+    ) -> Pin<
+        Box<
+            dyn Future<
+                    Output = Result<(Box<dyn AsyncReadWrite + Send + Unpin + 'static>, ListenAddr)>,
```

Comment:
Rather than returning a boxed IO resource, I would prefer to use an associated type.

(And if we _did_ want to return a boxed IO resource, it should have been a `Pin<Box<...>>` rather than an `Box<... + Unpin>`.)